### ISSUE TITLE
poetry won't check for pre-releases during dependency resolving of non-direct dependencies 

### ISSUE THREAD

**Issue by struegamer** [python-poetry/poetry#4405]
- [X] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [X] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [X] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **MacOS Catalina and also on Ubuntu Linux 18.04/20.04**: 
- **1.1.7**: 

## Issue

So, I have three python libs all of them are managed by poetry.

* lib-a
* lib-b
* lib-c

All libs are building fine and can be uploaded to a repository.

lib-b and lib-c are dependencies of lib-a, so far so good.

I changed the python version dependency of all three libs to 

```
[tool.poetry.dependencies]
python = "~3.6 || ~3.8"
```

And all lib versions are pre-release semver version (i.e. 0.1.2-alpha.0, etc.) so, there are no major.minor.patch versions for now.
The dep lines for lib-a to add lib-b and lib-c as deps are looking like this:

```
lib-b = {version = "^0.1.0", allow-prereleases = true}
lib-c = {version = "^0.1.0", allow-prereleases = true}
```
doing a `poetry install` with this, it all works smooth and nicely.

Now I created a tool which is using lib-a as direct dependency,

`pyproject.toml` looks like this:

```
[tool.poetry]
name = "awesome-project"
version = "0.1.5-alpha.0"
description = ""
authors = ["me <me@here.now>"]
packages = [
  {include="awesomeproject", from="src"}
]

[tool.poetry.dependencies]
python = "~3.6 || ~3.8"
lib-a = {version = "^1.0.0", allow-prereleases = true}
....
```
Now lib-a has a pre-release version number, lib-b and lib-c too. lib-a pulls in lib-b and lib-c and it also references `allow-prereleases=true`.

Expectation would be that `poetry install` would check all python version deps from lib-a and it would also check the python version deps from lib-b and lib-c and also takes `allow-prerelease=true` setting for these two libs in lib-a into account.

Sadly, it breaks, telling me that lib-a is not compatible with the `awesome-project` python version setting, because lib-a checks for lib-b and lib-c not for the pre-release versions, only for `major.minor.patch` versions. Which is obviously not available.

Now, to test that this assumption is true, I updated lib-b and lib-c version to a `major.minor.patch` version, uploaded them to my local pypi repo, `poetry update`d lib-a (still using pre-release version number) and uploaded this to my local pypi repo.

Now for `awesome-project` doing a `poetry install` and  it magically works now. all the same settings, just the version numbers of lib-b and lib-c changed.

Real bug is now: poetry won't check for pre-releases during dependency resolving of non-direct dependencies. Even when the version requirements are correctly set in the generated `setup.py` of the uploaded tarball.

I know dependency resolving is not easy, especially in python :) But during development you don't want to rely on the "latest stable" version of a library (direct or indirect), I guess there needs to be a fix for this issue.

**Comment by TBBle** [comment]
I think I just hit this in Poetry 1.2.0b2, in a trivial project (created with `poetry-1.2 init -n; poetry-1.2 lock`.
```
> poetry-1.2 --version
Poetry (version 1.2.0b2)
> poetry-1.2 add --lock --dry-run opentelemetry-distro
Using version ^0.31b0 for opentelemetry-distro

Updating dependencies
Resolving dependencies... (0.0s)

Because no versions of opentelemetry-distro match >0.31b0,<0.31
 and opentelemetry-distro (0.31b0) depends on opentelemetry-sdk (1.12.0rc1), opentelemetry-distro (>=0.31b0,<0.31) requires opentelemetry-sdk (1.12.0rc1).
And because opentelemetry-sdk (1.12.0rc1) depends on opentelemetry-api (1.12.0rc1), opentelemetry-distro (>=0.31b0,<0.31) requires opentelemetry-api (1.12.0rc1).
So, because no versions of opentelemetry-api match 1.12.0rc1
 and scratch depends on opentelemetry-distro (^0.31b0), version solving failed.
```

This is surprising because [`opentelemetry-distro`'s requirement is `opentelemetry-sdk == 1.12.0rc1`](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/v0.31b0/opentelemetry-distro/setup.cfg#L46), i.e. Poetry has _no_ discretion on which version is matched. As discussed in #3538, `--allow-prereleases` has no effect here because that only affects the generaion of constraints for entries on the command-line, and it's already selected the correct version of `opentelemetry-distro`.

**Comment by Seeefo** [comment]
I am running into this issue also. Specifically in a similar situation to @TBBle with Opentelemetry. Interally, we have a data generation package that uses `opentelemetry-sdk` directly (which has its own internal dependency on `opentelemetry-api`). To test some of the upcoming changes and how they would affect us, I swapped our dependency to `1.12.0rc2`. Testing the data generation package does fine (it uses pip). But some of its downstream dependencies use poetry and barf on `opentelemetry-api`:

Sum up:
- Package A depends on `opentelemetry-sdk 1.12.0rc2`
- Package B depends on Package A

Results:
Package B fails to resolve `opentelemetry-api` dependency from `opentelemetry-sdk`, even though that version does exist:
`Because no versions of otlp-data-generator match >0.0.1,<0.0.2
   and otlp-data-generator (0.0.1) depends on opentelemetry-sdk (1.12.0rc2), otlp-data-generator (>=0.0.1,<0.0.2) requires opentelemetry-sdk (1.12.0rc2).
  And because opentelemetry-sdk (1.12.0rc2) depends on opentelemetry-api (1.12.0rc2), otlp-data-generator (>=0.0.1,<0.0.2) requires opentelemetry-api (1.12.0rc2).
  So, because no versions of opentelemetry-api match 1.12.0rc2
   and storage-etl depends on otlp-data-generator (^0.0.1), version solving failed.`

To further test this, I pushed out a version of Package A that explicitly lists a dependency on `opentelemetry-api 1.12.0rc2`. The dependency failure in Package B has gone away.

I would expect these to resolve just fine, especially if `--allow-prereleases` is used...but I am not sure where the line would be drawn for resolving this.

**Comment by dimbleby** [comment]
I only see one reproducer in this thread, and it works fine today
```console
$ poetry init -n
$ poetry add --lock opentelemetry-distro==0.31b0

Updating dependencies
Resolving dependencies... (0.2s)

Writing lock file
$ poetry export --without-hashes
Warning: poetry-plugin-export will not be installed by default in a future version of Poetry.
In order to avoid a breaking change and make your automation forward-compatible, please install poetry-plugin-export explicitly. See https://python-poetry.org/docs/plugins/#using-plugins for details on how to install a plugin.
To disable this warning run 'poetry config warnings.export false'.
deprecated==1.2.14 ; python_version >= "3.10" and python_version < "4.0"
opentelemetry-api==1.12.0rc1 ; python_version >= "3.10" and python_version < "4.0"
opentelemetry-distro==0.31b0 ; python_version >= "3.10" and python_version < "4.0"
opentelemetry-instrumentation==0.31b0 ; python_version >= "3.10" and python_version < "4.0"
opentelemetry-sdk==1.12.0rc1 ; python_version >= "3.10" and python_version < "4.0"
opentelemetry-semantic-conventions==0.31b0 ; python_version >= "3.10" and python_version < "4.0"
setuptools==69.1.0 ; python_version >= "3.10" and python_version < "4.0"
typing-extensions==4.9.0 ; python_version >= "3.10" and python_version < "4.0"
wrapt==1.16.0 ; python_version >= "3.10" and python_version < "4.0"
```
NB various pre-releases resolved as expected

I guess this should be closed

**Comment by TBBle** [comment]
A quick poke at the history of the relevant file suggests #7978 was the fix, appearing in Poetry 1.5.1 and later.
